qurro/qarcoal.py

```diff
@@ -0,0 +1,113 @@
+    -----------
+        table: biom file with which to calculate log ratios
+        taxonomy: Qiime2 compliant feature taxonomy metadata file
+        num_string: numerator string to search for in taxonomy
+        denom_string: denominator string to search for in taxonomy
+        sample_to_use: Q2 Metadata file with samples to use (optional)
+
+    Returns:
+    --------
+        comparison_df: pd DataFrame in the form:
+
+            Sample-ID    Num_Sum    Denom_Sum   log_ratio
+                   S1          7           15   -0.762140
+    """
+
+    # biom table is features x samples
+    if samples_to_use is not None:
+        samp = set(samples_to_use.to_dataframe().index)
+        feat_table = table.filter(samp, axis="sample", inplace=False)
+        feat_table = feat_table.to_dataframe()
+    else:
+        feat_table = table.to_dataframe()
+
+    taxonomy_df = taxonomy.to_dataframe()
+
+    # taxonomy is features x [Taxon, Confidence]
+    taxonomy_df = taxonomy_df.loc[feat_table.index]
+    tax_num_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(num_string)]
+    tax_denom_df = taxonomy_df[taxonomy_df["Taxon"].str.contains(denom_string)]
+
+    if tax_num_df.shape[0] == 0:
+        if tax_denom_df.shape[0] == 0:
+            raise (ValueError("neither feature found!"))
+        else:
+            raise (ValueError("numerator not found!"))
+    elif tax_denom_df.shape[0] == 0:
+        raise (ValueError("denominator not found!"))
+    else:
+        pass
+
+    # drop columns (samples) in which no feature w/ string is present
+    tax_num_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+    tax_denom_df.drop(columns=["Taxon", "Confidence"], inplace=True)
+
+    tax_num_joined_df = tax_num_df.join(feat_table)
+    tax_num_joined_df = tax_num_joined_df.loc[
+        :, (tax_num_joined_df != 0).any(axis=0)
+    ]
+    tax_denom_joined_df = tax_denom_df.join(feat_table)
+    tax_denom_joined_df = tax_denom_joined_df.loc[
+        :, (tax_denom_joined_df != 0).any(axis=0)
+    ]
+
+    # keep only intersection of samples in which each feat string
+    # is present
+    samp_to_keep = set(tax_num_joined_df.columns).intersection(
+        set(tax_denom_joined_df.columns)
+    )
+    tax_num_joined_df = tax_num_joined_df[samp_to_keep]
+    tax_denom_joined_df = tax_denom_joined_df[samp_to_keep]
+
+    tax_num_sample_sum = tax_num_joined_df.sum(axis=0)
+    tax_denom_sample_sum = tax_denom_joined_df.sum(axis=0)
+
+    # sometimes 2 feature labels are the same i.e. same features
+    # don't want to check each time so only check if sums are
+    # the same
+    # e.g. if you are comparing g__A and s__B but g__A only
+    # appears when followed by s__B -> log ratios will all = 0
+    # TODO: Figure out if this makes sense at all lol
+    if tax_num_sample_sum.equals(tax_denom_sample_sum):
+        a = set(tax_num_df.index)
+        b = set(tax_num_df.index)
+        if a == b:
+            raise (ValueError("same features!"))
```

Comment:
I didn't really think about this until today :), but we're currently looking into warning the user if a selected log-ratio contains common features in the numerator and denominator (see #249), since having features shared in both sides is apparently not mathematically kosher.

It might be worth just checking for this in all cases, e.g.:

```python
if len(set(tax_num_df.index) & set(tax_denom_df.index)) > 0:
    raise ValueError("The selected log-ratio has features shared between the numerator and denominator!")
```

... and maybe adding an optional argument that'll let the user explicitly bypass this restriction.

Reply:
Got it, I can do that. 

Do you think the conditional in question (i.e. checking for the *exact* same features and returning an error) should be handled this way (if at all)? Something about it seems off (aside from the duplication of num referenced) but I can't put my finger on it...

Reply:
I don't think that the current conditional (`if tax_num_sample_sum.equals(tax_denom_sample_sum) ...`) should be handled this way, looking at it again. It's possible that you could have the numerator sums equal to the denominator sums for every sample, even if the actual features selected for the log-ratio are completely separate (this could happen in the silly, probs-never-gonna-happen case where e.g. ASV 1 and ASV 2 are always seen in equal relative abundance in every sample) -- this would just result in every sample having a log-ratio of 0.

If we're already going to warn the user about features being shared between the numerator and denominator, then I think it'd be safe to just remove this block of code. In the g__A and s__B example, if s__B is literally in the genus g__A then selecting the log-ratio of g__A / s__B would result in a warning/error about shared features; and if s__B isn't "in" g__A, then selecting g__A / s__B would be a completely valid thing to do and would would result in the all-0-log-ratios thing.

That was a big wall of text, so we can chat tomorrow about this if I wasn't clear here.

Reply:
Understood, I think this is a good solution.

Reply:
Ok, I've added a boolean option `allow_shared_features` that lets the user specify whether shared features should raise an error (defaults to False -> raise error).

Reply:
Sweet, thanks!